### PR TITLE
fix(campaigns): align LinkedIn pacing bands with shared thresholds

### PR DESCRIPTION
**Pull request by mrautela365**
## Summary

Part of [LFXV2-3314](https://linuxfoundation.atlassian.net/browse/LFXV2-3314) — the standalone pacing fix. **The rules migration itself is NOT in this PR; it is blocked (see below).**

`linkedin-ads.service.ts` hardcoded its own pacing bands (`40 / 90 / 105`) while `CAMPAIGN_PACING_THRESHOLDS` is `50 / 90 / 100`, and the action-item copy read *"Underspending — pacing below 40%"*.

The ticket recorded the decision: **50 is correct, the message is stale — fix the copy, not the threshold.** That decision assumed LinkedIn read the shared constant.

**It does not.** LinkedIn never imported `CAMPAIGN_PACING_THRESHOLDS` at all, so *"below 40%"* was accurate for the rule that actually fired. Editing only the copy would have made a true message false. The real defect is the divergence, which had two live consequences:

| | effect |
|---|---|
| **Bar vs label disagree** | `PacingClassPipe` colours the pacing bar off the shared `50`. A campaign at **45%** drew a red "underspending" bar beside a `normal` label — and raised **no action item**. |
| **Wrong band at the top** | Constrained ended at a local `105` vs the shared `100`, so **102%** read as `constrained` instead of `overspending`. |

This points the bands at the shared constants and interpolates both messages from them, so copy and rule cannot drift apart again. **Threshold values are unchanged** — the shared set is now simply the single source.

### Second commit: boundary parity (found in a pre-PR review sweep)

The first commit used a strict-less chain. Meta and Reddit test with `>` rather than `>=`, so a value equal to a threshold does not advance a band — each threshold is an **inclusive upper bound** of the band below it. **Those are not equivalent**, and they diverge at exactly two reachable values:

| pacingPct | strict-less | meta/reddit |
|---|---|---|
| **90** | constrained | normal |
| **100** | overspending | constrained |

Neither is an edge case — meta and reddit both `Math.round` `pacingPct`, so 90 and 100 are ordinary integer outcomes. The first commit would have left LinkedIn one band high at each, replacing the divergence it set out to remove with a narrower one. The second commit adopts the meta/reddit form so **all three platforms agree at every boundary**.

## Tests

New `linkedin-ads.service.spec.ts` — **7 tests, no prior spec existed** — driving the real `getLinkedInAnalytics` through a mocked fetch, with the flight pinned fully in the past so `pacingPct` is clock-independent.

Verified by mutation, not by passing:
- Reverting the threshold change fails **3** (the 45% label, the copy string, the 102% band).
- Reverting to the strict-less chain fails **exactly the 2** boundary cases (90, 100).

One test guards the LinkedIn-only "no ad creatives" rule so the pending migration cannot silently drop it.

`it()` counts: `linkedin-ads` 0 → 5 `it` blocks (one is `it.each` with 2 cases = 7 tests); `meta-ads` 14 → 14. **No tests removed or weakened.**

## Why the migration is not here

`GET /projects/{project_id}/briefs/{brief_id}/metrics` exists in campaign-service and returns `action_items` (LFXV2-3313 landed; confirmed in `design/brief.go` and `gen/http/openapi3.json`). **But nothing in the UI calls it** — no `getBriefMetrics` on `campaign-service.service.ts`, no `action_items` consumer outside the unrelated weekly-brief feature, and no `LFX_CUTOVER_CAMPAIGN_SERVICE_*` flag covering it.

Deleting the BFF rule implementations today would remove the Optimize tab's action items, not migrate them. The remaining work is **wiring**, not deletion.

## Known follow-up (not in this PR)

`reddit-ads.service.ts` still hardcodes `50 / 100 / 90` as literals. The values already match the shared constants, so there is no behavioural divergence — only the same drift risk. Left out to keep this PR to one capability; happy to sweep it here if preferred.

## Gates

`yarn format` (tree clean) · `yarn lint` · `yarn build` · `yarn test:app` · `yarn test:server` (73 files, 1717 tests) — all exit `0`. `git grep "DEV BYPASS"` empty. Commits are GPG-signed and signed-off.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01WFGQ59WQ2n8yLU4dGbMXnK

[LFXV2-3314]: https://linuxfoundation.atlassian.net/browse/LFXV2-3314?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by dealako — DISMISSED** [review]
Hi @mrautela365 👋 — re-reviewing at `533df181`. The description was already exemplary (it explains why the recorded "50 wins, fix the copy" decision rested on a false premise — LinkedIn never imported the shared constant — and lands on the route that keeps every message true). Pointing both the bands and the copy at `CAMPAIGN_PACING_THRESHOLDS`, the mutation-tested spec on a previously untested service, and preserving the LinkedIn-only "no ad creatives" rule ahead of the pending migration are all the right calls.

My first-round review surfaced one residual gap — the strict `<` chain still disagreed with `PacingClassPipe`'s inclusive `<=` at exactly 90% and 100% (same point Copilot flagged inline). Commit `533df181` ("match meta/reddit pacing band boundaries") landed while I was reviewing and resolves it thoroughly.

### Revision tracking — all prior findings resolved in `533df181`
1. ✅ **Boundary operator mismatch (was minor).** Switched to the meta/reddit exclusive-upper-bound form (`> constrained`, `> normal`). I verified label vs bar now agree at every boundary — 49/50/89/90/91/99/100/101/102 all match `PacingClassPipe`. At 90% both read normal; at 100% both read constrained. Resolves the Copilot inline comment too.
2. ✅ **Spec lacked boundary coverage (was minor).** Added table-driven cases pinning 90→normal and 100→constrained; reverting to the strict-less chain fails exactly those two.
3. ✅ **"pacing above 90%" copy at exactly 90% (was nit).** The `constrained` band now fires only at `pacingPct > 90`, so the copy is truthful at every value that renders it.

The commit message also documents *why* the comparison form is load-bearing (Meta and Reddit `Math.round` pacingPct, so 90 and 100 are ordinary integer outcomes, not edge cases) — good reasoning to capture in-tree.

### AI bot reconciliation
- **Copilot** flagged the boundary/operator issue inline — now resolved by `533df181`.
- **Cursor Bugbot** posted a general "Medium Risk" summary, no specific defect.
- **CodeRabbit** auto-review disabled on this repo.

### Security & Privacy
No findings. The diff interpolates only trusted numeric constants into response strings, adds no PII-carrying log/error/response paths, and account IDs remain masked. The spec's `LINKEDIN_ACCESS_TOKEN='test-token'` is a synthetic fixture.

**Final decision: ✅ Approved.** Threshold values unchanged, all three platforms now agree at every boundary, and the regression is pinned by tests. Nice follow-through on the pre-PR sweep.

File: apps/lfx-one/src/server/services/linkedin-ads.service.ts [hunk 1]
```diff
@@ -1010,11 +1010,15 @@ export async function getLinkedInAnalytics(req: Request | undefined, accountId:
...
+      // in `PacingClassPipe` colours the same campaign off CAMPAIGN_PACING_THRESHOLDS, so local
+      // numbers here made the label and the bar disagree (e.g. 45% read "underspending" red in
+      // the bar but "normal" in the label).
+      if (pacingPct < CAMPAIGN_PACING_THRESHOLDS.underspending) {
         pacingLabel = 'underspending';
-      } else if (pacingPct < 90) {
+      } else if (pacingPct < CAMPAIGN_PACING_THRESHOLDS.normal) {
         pacingLabel = 'normal';
-      } else if (pacingPct < 105) {
+      } else if (pacingPct < CAMPAIGN_PACING_THRESHOLDS.constrained) {
```
[the remark this answers is not in the document]
Comment (mrautela365) [reply apps/lfx-one/src/server/services/linkedin-ads.service.ts]: Agreed, and already fixed — thank you both, this is exactly right.

The finding is against `102b423ab`. A pre-PR review sweep caught the same divergence independently and `533df1811` landed the fix before your review posted, so the branch head no longer has it.

LinkedIn now uses the same comparison form as `meta-ads.service.ts` and `reddit-ads.service.ts`:

```ts
if (pacingPct < CAMPAIGN_PACING_THRESHOLDS.underspending) pacingLabel = 'underspending';
else if (pacingPct > CAMPAIGN_PACING_THRESHOLDS.constrained) pacingLabel = 'overspending';
else if (pacingPct > CAMPAIGN_PACING_THRESHOLDS.normal) pacingLabel = 'constrained';
```

All three platforms now agree at every boundary (90 → `normal`, 100 → `constrained`), and neither value is an edge case since Meta and Reddit both `Math.round` `pacingPct`.

On the secondary point about the copy at 90%: correct, and it was the half the label assertion did not cover. `Budget constrained — pacing above 90%` fires for `constrained` and `overspending` alike, so mislabelling exactly 90% told the operator it paced *above* 90% when it landed exactly on it. `ff99f3b2b` adds a case asserting no such item is emitted at exactly 90%.

Coverage is mutation-verified rather than merely green: reverting to the strict-less chain fails three cases (90 label, 100 label, and the 90% copy).

[the remark this answers is not in the document]
Comment (mrautela365) [reply apps/lfx-one/src/server/services/linkedin-ads.service.ts]: Agreed, and already fixed — thank you both, this is exactly right.

The finding is against `102b423ab`. A pre-PR review sweep caught the same divergence independently and `533df1811` landed the fix before your review posted, so the branch head no longer has it.

LinkedIn now uses the same comparison form as `meta-ads.service.ts` and `reddit-ads.service.ts`:

```ts
if (pacingPct < CAMPAIGN_PACING_THRESHOLDS.underspending) pacingLabel = 'underspending';
else if (pacingPct > CAMPAIGN_PACING_THRESHOLDS.constrained) pacingLabel = 'overspending';
else if (pacingPct > CAMPAIGN_PACING_THRESHOLDS.normal) pacingLabel = 'constrained';
```

All three platforms now agree at every boundary (90 → `normal`, 100 → `constrained`), and neither value is an edge case since Meta and Reddit both `Math.round` `pacingPct`.

On the secondary point about the copy at 90%: correct, and it was the half the label assertion did not cover. `Budget constrained — pacing above 90%` fires for `constrained` and `overspending` alike, so mislabelling exactly 90% told the operator it paced *above* 90% when it landed exactly on it. `ff99f3b2b` adds a case asserting no such item is emitted at exactly 90%.

Coverage is mutation-verified rather than merely green: reverting to the strict-less chain fails three cases (90 label, 100 label, and the 90% copy).

[the remark this answers is not in the document]
Comment (mrautela365) [reply apps/lfx-one/src/server/services/linkedin-ads.service.ts]: Respectfully disagreeing with this one — the finding rests on a premise the code does not match, and I want to leave the evidence checkable rather than just assert it.

**The bands do not use `<`.** At branch head `0c011132`, `linkedin-ads.service.ts:1024-1029`:

```ts
if (pacingPct < CAMPAIGN_PACING_THRESHOLDS.underspending) {
  pacingLabel = 'underspending';
} else if (pacingPct > CAMPAIGN_PACING_THRESHOLDS.constrained) {
  pacingLabel = 'overspending';
} else if (pacingPct > CAMPAIGN_PACING_THRESHOLDS.normal) {
  pacingLabel = 'constrained';
}
```

and `meta-ads.service.ts:539-541`:

```ts
if (pacingPct < CAMPAIGN_PACING_THRESHOLDS.underspending) pacingLabel = 'underspending';
else if (pacingPct > CAMPAIGN_PACING_THRESHOLDS.constrained) pacingLabel = 'overspending';
else if (pacingPct > CAMPAIGN_PACING_THRESHOLDS.normal) pacingLabel = 'constrained';
```

Same three comparisons in the same order. `reddit-ads.service.ts:286-288` is the same shape with literals. The only `<` is on `underspending`, which is a lower bound and is not what the finding is about.

So `90` and `100` already land where you say they should: at 90, `90 > 100` is false and `90 > 90` is false → `normal`. At 100, `100 > 100` is false and `100 > 90` is true → `constrained`. Pinned by tests at `linkedin-ads.service.spec.ts:166-171`, and mutation-verified — swapping in a strict-less chain fails exactly those two cases plus the 90% copy assertion.

This was real on `102b423ab`, which is the commit the review ran against. `533df181` fixed it before the review posted, so the thread is reporting genuinely stale code. Changing the logic now would reintroduce the bug.

**That said, the report was not unreasonable — the comment above the code was wrong**, and I think it is why the code read as `<`. It said "EXCLUSIVE upper bounds" when the behaviour is the opposite. Your other two threads caught that, and `0c011132` fixes all three sites. Marking this one resolved as a false positive on current head; the terminology fix is the real remedy.

File: apps/lfx-one/src/server/services/linkedin-ads.service.ts [hunk 2]
```diff
@@ -1010,14 +1010,23 @@ export async function getLinkedInAnalytics(req: Request | undefined, accountId:
...
     let pacingLabel: LinkedInPacingLabel = 'normal';
     if (hasBudget) {
-      if (pacingPct < 40) {
+      // Bands come from the shared constants rather than LinkedIn-local literals: the pacing bar
+      // in `PacingClassPipe` colours the same campaign off CAMPAIGN_PACING_THRESHOLDS, so local
+      // numbers here made the label and the bar disagree (e.g. 45% read "underspending" red in
+      // the bar but "normal" in the label).
+      //
+      // The comparison form deliberately mirrors meta-ads.service.ts and reddit-ads.service.ts
+      // exactly. Both treat the thresholds as EXCLUSIVE upper bounds (`> normal`, `> constrained`),
```
[the remark this answers is not in the document]
Comment (mrautela365) [reply apps/lfx-one/src/server/services/linkedin-ads.service.ts]: Correct, and thank you — fixed in `0c011132`.

You are right on the semantics: with `>` rather than `>=`, a value EQUAL to the threshold does not advance a band, so the threshold is an **inclusive upper bound** of the band below it (equivalently an exclusive lower bound of the next). I had it backwards.

The comment now reads:

> Because the tests are `>` rather than `>=`, a value EQUAL to a threshold does not advance a band: each threshold is an INCLUSIVE upper bound of the band below it, so 90% stays `normal` and 100% stays `constrained`.

Worth flagging that this was not cosmetic: the wrong word is almost certainly why the other thread on this PR read the code as using `<` and filed a boundary bug against logic that is character-for-character identical to `meta-ads.service.ts:539-541`. A comment describing the opposite of its code cost a review round here, which is a good argument for treating that class as a real defect rather than a nit.

I swept the term rather than fixing only the flagged lines — three sites, all introduced by this branch: this comment, the spec block comment at `:160`, and the `it.each` title at `:168`. No other occurrence in `apps/` or `packages/` (the remaining "exclusive" matches are unrelated, e.g. mutually-exclusive params). Threshold values and assertions untouched; `it` count stays 6 + 1 `it.each` = 8 tests, and reverting the band form still fails exactly 3.

File: apps/lfx-one/src/server/services/linkedin-ads.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+  // Boundary parity with meta-ads.service.ts / reddit-ads.service.ts, which both treat the
+  // thresholds as EXCLUSIVE upper bounds. Both also `Math.round` pacingPct, so exactly 90 and
+  // exactly 100 are ordinary integer outcomes rather than edge cases. A strict-less chain
+  // (`pacingPct < normal` / `< constrained`) puts each of these one band too high, which is what
+  // the first draft of this change did.
+  it.each([
+    [90, 'normal'],
+    [100, 'constrained'],
+  ])('matches meta/reddit at exactly %i%% (a threshold is an exclusive upper bound)', async (pct, expected) => {
```
[the remark this answers is not in the document]
Comment (mrautela365) [reply apps/lfx-one/src/server/services/linkedin-ads.service.spec.ts]: Agreed — renamed in `0c011132`.

The title now reads *"matches meta/reddit at exactly %i%% (a threshold is an inclusive upper bound)"*, which is what the assertions actually establish: 90 → `normal`, 100 → `constrained`.

Same sweep as the sibling thread — the block comment directly above this case had the identical error and is fixed, as is the source comment in `linkedin-ads.service.ts` that this title was mirroring.

Rename only: no assertion changed, `it` count unchanged at 6 + 1 `it.each` (8 tests), and the cases still bind — reverting the band form to a strict-less chain fails exactly these two plus the 90% copy assertion.

**Comment by dealako** [comment]
Hi @mrautela365 👋 — follow-up review at `0c011132`. Thanks for the quick, precise turnaround on the boundary-semantics thread.

### 👏 Nice work this round
- **Terminology corrected the right direction.** The service comment and the `it.each` title now say the thresholds are *inclusive* upper bounds — which is what the `> normal` / `> constrained` chain actually implements (90 stays `normal`, 100 stays `constrained`). This matches Copilot's two comments exactly, and matters because the *next* person to touch these bands reads that comment first.
- **Pinned the copy half of the bug.** The new test `emits no "pacing above 90%" item for a campaign sitting exactly on 90%` guards the operator-facing string directly, not just the label. That's the half that's easiest to regress silently, since the "Budget constrained — pacing above 90%" copy is shared by both `constrained` and `overspending`.
- **Boundary cases held.** The `it.each` still pins 90→`normal` and 100→`constrained`, so all three platforms (LinkedIn/Meta/Reddit) and `PacingClassPipe` agree at every threshold.

### Revision tracking — all prior items resolved
- ✅ **P1** (label vs bar diverged at 90/100): resolved in `533df181`, unchanged here. Verified logic matches `PacingClassPipe` at every boundary.
- ✅ **P2** (spec lacked 90/100 boundary tests): resolved; both `it.each` rows present and asserting.
- ✅ **P3** (`"pacing above 90%"` copy at exactly 90%): resolved and now regression-pinned by the new test.
- ✅ **Copilot: service comment terminology** (EXCLUSIVE→INCLUSIVE): fixed in `0c011132`.
- ✅ **Copilot: test title terminology**: fixed in `0c011132`.

### Issue count (carried over + new)
- 🔴 Blocking: 0
- 🟡 Minor: 0
- ⚪ Nit: 0

### AI bot reconciliation
- **Copilot** raised the two terminology comments this round — both addressed in `0c011132`, and Copilot's latest re-review generated no new comments.
- **Cursor Bugbot** / **CodeRabbit**: nothing new to reconcile.

### Security & Privacy
No findings. These two commits are comment- and test-only; no new PII, secrets, or logic surface.

**Final decision: ✅ Approved.** Clean follow-through — semantics, tests, and docs all line up now.

**Review by dealako — APPROVED** [review]
Follow-up review — approved. All prior feedback and both Copilot terminology comments resolved at 0c011132. See the summary comment for details.
